contracts/upgradeable_contracts/GasTokenConnector.sol

```diff
@@ -0,0 +1,96 @@
+    bytes4 internal constant ON_TOKEN_TRANSFER = 0xa4c0ed36; // onTokenTransfer(address,uint256,bytes)
+
+    /**
+    * @dev Updates gas token related parameters, target mint value and token receiver
+    * @param _targetMintValue target amount of token to mint/transfer during one call to _collectGasTokens
+    * @param _receiver receiver of minted gas token
+    */
+    function setGasTokenParameters(uint256 _targetMintValue, address _receiver) external onlyOwner {
+        uintStorage[GAS_TOKEN_TARGET_MINT_VALUE] = _targetMintValue;
+        addressStorage[GAS_TOKEN_RECEIVER] = _receiver;
+    }
+
+    /**
+    * @dev Updates gas token target mint value
+    * @param _targetMintValue target amount of token to mint/transfer during one call to _collectGasTokens
+    */
+    function setGasTokenTargetMintValue(uint256 _targetMintValue) external onlyOwner {
+        uintStorage[GAS_TOKEN_TARGET_MINT_VALUE] = _targetMintValue;
+    }
+
+    /**
+    * @dev Updates gas token receiver
+    * @param _receiver receiver of minted gas token
+    */
+    function setGasTokenReceiver(address _receiver) external onlyOwner {
+        addressStorage[GAS_TOKEN_RECEIVER] = _receiver;
+    }
+
+    /**
+    * @return Configured gas token target mint value
+    */
+    function gasTokenTargetMintValue() public view returns (uint256) {
+        return uintStorage[GAS_TOKEN_TARGET_MINT_VALUE];
+    }
+
+    /**
+    * @return Configured gas token receiver
+    */
+    function gasTokenReceiver() public view returns (address) {
+        return addressStorage[GAS_TOKEN_RECEIVER];
+    }
+
+    /**
+    * @return GST2 contract address in ethereum Mainnet
+    */
+    function gasToken() public pure returns (IGasToken) {
+        return IGasToken(0x0000000000b3F879cb30FE243b4Dfee438691c04);
+    }
+
+    /**
+    * @dev Internal function, for collection gas tokens
+    * If sufficient allowance from sender exist, the new tokens are not minted at all
+    * Otherwise, all allowance is taken + remaining tokens are minted
+    * In total, configured target amount of tokens is transfered to gas token receiver
+    */
+    function _collectGasTokens() internal {
+        uint256 target = gasTokenTargetMintValue();
+
+        if (target == 0) return;
+
+        address receiver = gasTokenReceiver();
+        if (receiver == address(0)) return;
+
+        IGasToken gst = gasToken();
+        uint256 approved = gst.allowance(msg.sender, address(this));
+
+        if (approved >= target) {
+            // tokens are transfered directly to receiver, avoiding transfer to the bridge contract
+            require(gst.transferFrom(msg.sender, receiver, target));
+        } else {
+            if (approved > 0) {
+                require(gst.transferFrom(msg.sender, address(this), approved));
+            }
+
+            gst.mint(target - approved);
```

Comment:
Could you provide more details for the following case:
1. Imagine that the bridge contract invoker minted gas token for the same gas price as the current transaction is being executed.
2. What will be evidently cheaper:
    - mint and approve the gas token in the invoker just before calling `requireToPassMessage`?
    - mint tokens and transfer them to the invoker in one transaction and trigger the invoker in another transaction?
    - do nothing as so the bridge contract will mint gas token by itself?

Reply:
So, I've created a simple invoker for tests, the results are the following, assuming bridge requires `0.5 GST2` to provers a request:
* invoker just calls `requireToPassMessage`, bridge does all minting work: `1884740` gas
* invoker mints tokens by himself, then approves them and calls `requireToPassMessage`: `1893440` gas
* someone else mints gas tokens, then transfers it to the invoker, then invoker calls `requireToPassMessage`: `1860634` + `51759` + `45790` = `1958183` gas

In general, just calling `requireToPassMessage` is the simplest/cheapest way. But for more complex scenarios, where caller has tight bounds on gas limit for `requireToPassMessage` call or already have some gas tokens, the approach `approve + requireToPassMessage` in a separate transaction should be preferred.

Reply:
Great research! Thanks! Could you create a page under `https://docs.tokenbridge.net/amb-bridge/` where describe the main idea about the gas token usage in AMB and describe these scenarios and the scenario with `tx.origin`?

Reply:
Yes, sure. I will start to write it today